homeassistant/components/zha/__init__.py

```diff
@@ -132,19 +170,48 @@ class RadioType(enum.Enum):
     hass.services.async_register(DOMAIN, SERVICE_REMOVE, remove,
                                  schema=SERVICE_SCHEMAS[SERVICE_REMOVE])
 
+    def zha_shutdown(event):
+        """Close radio."""
+        hass.data[DATA_ZHA][DATA_ZHA_RADIO].close()
+
+    hass.bus.async_listen_once(ha_const.EVENT_HOMEASSISTANT_STOP, zha_shutdown)
+    return True
+
+
+async def async_unload_entry(hass, config_entry):
+    """Unload ZHA config entry."""
+    hass.services.async_remove(DOMAIN, SERVICE_PERMIT)
+    hass.services.async_remove(DOMAIN, SERVICE_REMOVE)
+
+    for component in COMPONENTS:
+        await hass.config_entries.async_forward_entry_unload(
+            config_entry, component)
+
+    dispatchers = hass.data[DATA_ZHA].get(DATA_ZHA_DISPATCHERS, [])
+    for unsub_dispatcher in dispatchers:
+        unsub_dispatcher()
```

Comment:
It follows a better logic to unsubscribe first then do the unloading of the platforms.